# Предобработка данных

Отработка алгоритма решения задачи предобработки данных. 

Датасет <code>dtp_last.csv</code> содержит информацию о ДТП в Российской Федерации в 2018 году.

<li>'reg_code' - код региона (октмо)
<li>'регион' - название региона
<li>'road_code' - код дороги
<li>'Дорога' -  название дороги
<li>'road_type' - тип дороги
<li>'address' -  адрес
<li>'crash_type_name' -тип ДТП
<li>'crash_date' - дата ДТП
<li>'crash_time' - время ДТП
<li>'crash reason' - причина ДТП
<li>'fatalities_amount' - число погибших
<li>'victims_amount' - число пострадавших
<li>'vehicles_amount' - число транспортных средств
<li>'participants_amount' - число участников



Задача: Подготовить данные для дальнейшего исследования.
1. Изучите общую информацию.
2. Осуществите предобработку заголовков.
3. Заполните пропуски.
4. Найдите проблемы в данных.
5. Преобразуйте типы данных. Добавьте столбец datetime из имеющихся столбцов, используя разделитель T. Преобразуйте к типу данных "дата/время".
6. Удалите дубликаты.
7. Добавьте столбец с днями недели на русском языке.
8. Экспортируйте результат в excel.
9. Сформулируйте общий вывод по работе.

Работу сопроводите выводами и комментариями к коду.

### Загрузка данных. Общая информация

In [177]:
import pandas as pd
df = pd.read_csv('dtp_last.csv', sep = ';')
df.head(40)

reg_code                          регион    road_code  \
0         46              Московская область          NaN   
1         53            Оренбургская область  53К-2118000   
2         46              Московская область        А-107   
3          8                Хабаровский край        А-375   
4         35                 Республика Крым      35Н-076   
5         26            Республика Ингушетия        Р-217   
6          1                  Алтайский край          NaN   
7         80         Республика Башкортостан          М-7   
8         56              Пензенская область          NaN   
9         35                 Республика Крым      35К-023   
10        35                 Республика Крым      35К-015   
11        79               Республика Адыгея          М-4   
12         7             Ставропольский край          NaN   
13         3              Краснодарский край          NaN   
14        93                 Республика Тыва          NaN   
15        57                   Пермский край        Р-242   
16        97  Чувашская Республика — Чувашия          NaN   
17        30                 Камчатский край          NaN   
18        25               Иркутская область          NaN   
19        50           Новосибирская область          NaN   
20        30                 Камчатский край          NaN   
21        29               Калужская область          NaN   
22        33               Кировская область          NaN   
23        99    Еврейская автономная область          NaN   
24        11           Архангельская область      11Р-001   
25        70                Тульская область          NaN   
26        17            Владимирская область          М-7   
27        80         Республика Башкортостан          NaN   
28         3              Краснодарский край          М-4   
29        22           Нижегородская область     22К-0016   
30        79               Республика Адыгея        А-159   
31        79               Республика Адыгея        А-146   
32        12            Астраханская область         Р-22   
33        56              Пензенская область          NaN   
34         3              Краснодарский край        А-147   
35         3              Краснодарский край          NaN   
36        24              Ивановская область          NaN   
37         7             Ставропольский край        Р-217   
38        46              Московская область          NaN   
39        57                   Пермский край          NaN   

                                               Дорога  \
0                        мкр. Авиационный-Лямцино-ММК   
1                                  Обход г. Оренбурга   
2   "Московское малое кольцо" Икша - Ногинск - Бро...   
3   "Восток" Хабаровск - Красный Яр - Ариадное - Ч...   
4              Симферополь - Севастополь - Бахчисарай   
5   "Кавказ" автомобильная дорога М-4 "Дон" - Влад...   
6                             Вторая Каменка - Гилево   
7   "Волга" Москва - Владимир - Нижний Новгород - ...   
8   "Автодорога ""р.п. Беково - с. Варварино - р.п...   
9                      Восточный обход г. Симферополя   
10                             Раздольное - Евпатория   
11  "Дон" Москва - Воронеж - Ростов-на-Дону - Крас...   
12  Ачикулак - Махмуд-Мектеб Тукуй Мектеб 0000 - 4...   
13               г.Армавир-с.Успенское-г.Невинномысск   
14                             Подъезд к с. Кара-Хаак   
15                               Пермь - Екатеринбург   
16                 "Аликово-Старые Атаи-а/д ""Сура"""   
17                        г.П-Камчатский - с.Мильково   
18                                      Иркутск-Качуг   
19   К-17р Новосибирск - Кочки - Павлодар (в пред.РФ)   
20  с/х Начикинский-п.Усть-Большерецк-п.Октябрьски...   
21  Калуга - Ферзиково - Таруса - Серпухов - Авчурино   
22                           Киров - Советск - Яранск   
23  Подъезд к международному речному порту в с. Н....   
24  Архангельск (от пос. Брин-Наволок) Каргополь В...   
25                  

In [178]:
df.shape

(39102, 14)

In [179]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39102 entries, 0 to 39101
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   reg_code             39102 non-null  int64 
 1   регион               39083 non-null  object
 2   road_code            20777 non-null  object
 3   Дорога               38910 non-null  object
 4   road_type            38910 non-null  object
 5   address              39102 non-null  object
 6   crash_type_name      39102 non-null  object
 7   crash_date           39102 non-null  int64 
 8   crash_time           39102 non-null  object
 9   crash reason         38467 non-null  object
 10  fatalities_amount    39102 non-null  int64 
 11  victims_amount       39102 non-null  int64 
 12  vehicles_amount      39102 non-null  int64 
 13  participants_amount  39102 non-null  object
dtypes: int64(5), object(9)
memory usage: 4.2+ MB


### Выводы:

1. датасет содержит 39102 строки и 14 столбцов
2. возможно, коды (ОКТМО) не соответсвуют названиям регионов, надо проверять
3. в некоторых стобцах, например "road_code", есть отсутсвующие значения
4. необходимо поменять тип данных столбца "crash_date"
5. заголовки необходимо сделать на одном языке

### Предобработка заголовков

In [180]:
df = df.rename(columns={
    'reg_code': 'Код региона',
    'регион': 'Регион',
    'road_code': 'Код дороги',
    'Дорога': 'Название дороги',
    'road_type': 'Тип дороги',
    'address': 'Адрес',
    'crash_type_name': 'Тип ДТП',
    'crash_date': 'Дата ДТП',
    'crash_time': 'Время ДТП',
    'crash_reason': 'Причина ДТП',
    'crash reason': 'Причина ДТП',
    'fatalities_amount': 'Число погибших',
    'victims_amount': 'Число пострадавших',
    'vehicles_amount': 'Число транспортных средств',
    'participants_amount': 'Число участников'
})

In [181]:
df.head()

Код региона                Регион   Код дороги  \
0           46    Московская область          NaN   
1           53  Оренбургская область  53К-2118000   
2           46    Московская область        А-107   
3            8      Хабаровский край        А-375   
4           35       Республика Крым      35Н-076   

                                     Название дороги  \
0                       мкр. Авиационный-Лямцино-ММК   
1                                 Обход г. Оренбурга   
2  "Московское малое кольцо" Икша - Ногинск - Бро...   
3  "Восток" Хабаровск - Красный Яр - Ариадное - Ч...   
4             Симферополь - Севастополь - Бахчисарай   

                                          Тип дороги  \
0  Региональная или межмуниципальная дорога общег...   
1  Региональная или межмуниципальная дорога общег...   
2              Федеральная дорога общего пользования   
3              Федеральная дорога общего пользования   
4  Региональная или межмуниципальная дорога общег...   

                                               Адрес               Тип ДТП  \
0  ЦФО, Московская область, Городской округ Домод...         Опрокидывание   
1                ПФО, Оренбургская область, Оренбург  Наезд на препятствие   
2         ЦФО, Московская область, Дмитровский район         Опрокидывание   
3           ДФО, Хабаровский край, Хабаровский район     Наезд на пешехода   
4         ЮФО, Республика Крым, Бахчисарайский район     Наезд на пешехода   

   Дата ДТП Время ДТП                                        Причина ДТП  \
0  20181231  23:55:00                    Другие нарушения ПДД водителями   
1  20181231  23:55:00                    Другие нарушения ПДД водителями   
2  20181231  23:40:00         Превышение установленной скорости движения   
3  20181231  23:20:00  Нахождение на проезжей части без цели её перех...   
4  20181231  23:00:00      Нарушение правил проезда пешеходного перехода   

   Число погибших  Число пострадавших  Число транспортных средств  \
0               1                   0                           1   
1               1                   0                           1   
2               1                   0                           1   
3               0                   1                           1   
4               1                   0                           1   

  Число участников  
0               12  
1                2  
2                2  
3                2  
4                2

### Обработка пропущенных значений

In [182]:
print(df.isnull().sum()) #проверка наличия пропущенных значений

Код региона                       0
Регион                           19
Код дороги                    18325
Название дороги                 192
Тип дороги                      192
Адрес                             0
Тип ДТП                           0
Дата ДТП                          0
Время ДТП                         0
Причина ДТП                     635
Число погибших                    0
Число пострадавших                0
Число транспортных средств        0
Число участников                  0
dtype: int64


In [183]:
# заполнение пропущенных значений в столбцах нулями или информацией об отсутствии данных
for col in df.columns:
    if df[col].dtype == 'int64':
        df[col].fillna(0, inplace=True)
    elif df[col].dtype == 'object':
        df[col].fillna('Нет данных', inplace=True)
# проверка наличия пропущенных значений после заполнения
print(df.isnull().sum())

Код региона                   0
Регион                        0
Код дороги                    0
Название дороги               0
Тип дороги                    0
Адрес                         0
Тип ДТП                       0
Дата ДТП                      0
Время ДТП                     0
Причина ДТП                   0
Число погибших                0
Число пострадавших            0
Число транспортных средств    0
Число участников              0
dtype: int64


### Поиск проблем в данных

In [184]:
#список кодов ОКТМО
region_table = [
    (1, 'Алтайский край'),
    (3, 'Краснодарский край'),
    (4, 'Красноярский край'),
    (5, 'Приморский край'),
    (7, 'Ставропольский край'),
    (8, 'Хабаровский край'),
    (10, 'Амурская область'),
    (11, 'Архангельская область'),
    (12, 'Астраханская область'),
    (14, 'Белгородская область'),
    (15, 'Брянская область'),
    (17, 'Владимирская область'),
    (18, 'Волгоградская область'),
    (19, 'Вологодская область'),
    (20, 'Воронежская область'),
    (22, 'Нижегородская область'),
    (24, 'Ивановская область'),
    (25, 'Иркутская область'),
    (26, 'Республика Ингушетия'),
    (27, 'Калининградская область'),
    (28, 'Тверская область'),
    (29, 'Калужская область'),
    (30, 'Камчатская область'),
    (32, 'Кемеровская область'),
    (33, 'Кировская область'),
    (34, 'Костромская область'),
    (35, 'Республика Крым'),
    (36, 'Самарская область'),
    (37, 'Курганская область'),
    (38, 'Курская область'),
    (40, 'Город Санкт-Петербург'),
    (41, 'Ленинградская область'),
    (42, 'Липецкая область'),
    (44, 'Магаданская область'),
    (45, 'Город Москва'),
    (46, 'Московская область'),
    (47, 'Мурманская область'),
    (49, 'Новгородская область'),
    (50, 'Новосибирская область'),
    (52, 'Омская область'),
    (53, 'Оренбургская область'),
    (54, 'Орловская область'),
    (56, 'Пензенская область'),
    (57, 'Пермский край'),
    (58, 'Псковская область'),
    (60, 'Ростовская область'),
    (61, 'Рязанская область'),
    (63, 'Саратовская область'),
    (64, 'Сахалинская область'),
    (65, 'Свердловская область'),
    (66, 'Смоленская область'),
    (67, 'Город Севастополь'),
    (68, 'Тамбовская область'),
    (69, 'Томская область'),
    (70, 'Тульская область'),
    (71, 'Тюменская область'),
    (73, 'Ульяновская область'),
    (75, 'Челябинская область'),
    (76, 'Забайкальский край'),
    (77, 'Чукотский автономный округ'),
    (78, 'Ярославская область'),
    (79, 'Республика Адыгея (Адыгея)'),
    (80, 'Республика Башкортостан'),
    (81, 'Республика Бурятия'),
    (82, 'Республика Дагестан'),
    (83, 'Кабардино-Балкарская Р-ка'),
    (84, 'Республика Алтай'),
    (85, 'Республика Калмыкия'),
    (86, 'Республика Карелия'),
    (87, 'Республика Коми'),
    (88, 'Республика Марий Эл'),
    (89, 'Республика Мордовия'),
    (90, 'Р-ка Северная Осетия-Алания'),
    (91, 'Карачаево-Черкесская Р-ка'),
    (92, 'Республика Татарстан'),
    (93, 'Республика Тыва'),
    (94, 'Удмуртская Республика'),
    (95, 'Республика Хакасия'),
    (96, 'Чеченская Республика'),
    (97, 'Чувашская Республика'),
    (98, 'Республика Саха (Якутия)'),
    (99, 'Еврейская автономная область'),
    (718, 'Ханты-Мансийский автономный округ — Югра'),
    (719, 'Ямало-Ненецкий автономный округ')
]

In [185]:
def checkRegion(row):
    for region in region_table:
        key = region[0]
        value = region[1]
        if row['Код региона'] == key:
            # if row['Регион'] != value:
            # print(row['Код региона'], row['Регион'], value) #- здесь можно посмотреть что было изменено
            return value
    print(row['Код региона'], row['Регион'])  # - здесь можно посмотреть что не нашлось в таблице.
    # 718 Ханты-Мансийский автономный округ — Югра
    # 719 Ямало-Ненецкий автономный округ
    # видим что в таблице нет этих регионов, добавляем их
    return "Отсутствует в таблице"


df['Регион'] = df.apply(checkRegion, axis=1)
# проблема с 718 и 719. таких регионов нет в таблице
# print(df[df['Регион'] == 'Отсутствует в таблице']) # пусто, так как все нашлось

In [186]:
df

Код региона                      Регион   Код дороги  \
0               46          Московская область   Нет данных   
1               53        Оренбургская область  53К-2118000   
2               46          Московская область        А-107   
3                8            Хабаровский край        А-375   
4               35             Республика Крым      35Н-076   
...            ...                         ...          ...   
39097           80     Республика Башкортостан      80Н-069   
39098           79  Республика Адыгея (Адыгея)   Нет данных   
39099            3          Краснодарский край        Р-217   
39100           76          Забайкальский край        А-350   
39101           63         Саратовская область   Нет данных   

                                         Название дороги  \
0                           мкр. Авиационный-Лямцино-ММК   
1                                     Обход г. Оренбурга   
2      "Московское малое кольцо" Икша - Ногинск - Бро...   
3      "Восток" Хабаровск - Красный Яр - Ариадное - Ч...   
4                 Симферополь - Севастополь - Бахчисарай   
...                                                  ...   
39097                   Аскарово - Альмухаметово - Сибай   
39098                               Подъезд к а.Вочепший   
39099               Подъезд к городу Майкоп от а/д Р-217   
39100  Чита - Забайкальск - граница с Китайской Народ...   
39101  "автомобильная дорога ""Самара - Пугачев - Энг...   

                                              Тип дороги  \
0      Региональная или межмуниципальная дорога общег...   
1      Региональная или межмуниципальная дорога общег...   
2                  Федеральная дорога общего пользования   
3                  Федеральная дорога общего пользования   
4      Региональная или межмуниципальная дорога общег...   
...                                                  ...   
39097  Региональная или межмуниципальная дорога общег...   
39098  Региональная или межмуниципальная дорога общег...   
39099              Федеральная дорога общего пользования   
39100              Федеральная дорога общего пользования   
39101  Региональная или межмуниципальная дорога общег...   

                                                   Адрес  \
0      ЦФО, Московская область, Городской округ Домод...   
1                    ПФО, Оренбургская область, Оренбург   
2             ЦФО, Московская область, Дмитровский район   
3               ДФО, Хабаровский край, Хабаровский район   
4             ЮФО, Республика Крым, Бахчисарайский район   
...                                                  ...   
39097  ПФО, Республика Башкортостан, Абзелиловский ра...   
39098           ЮФО, Республика Адыгея, Теучежский район   
39099       ЮФО, Краснодарский край, Новокубанский район   
39100  СФО, Забайкальский край, Читинский район, Ново...   
39101        ПФО, Саратовская область, Энгельсский район   

                    Тип ДТП  Дата ДТП Время ДТП  \
0             Опрокидывание  20181231  23:55:00   
1      Наезд на препятствие  20181231  23:55:00   
2             Опрокидывание  20181231  23:40:00   
3         Наезд на пешехода  20181231  23:20:00   
4         Наезд на пешехода  20181231  23:00:00   
...                     ...       ...       ...   
39097     Наезд на животное  20180401   1:55:00   
39098     Наезд на животное  20180401   1:30:00   
39099          Столкновение  20180401   1:00:00   
39100         Опрокидывание  20180401   0:30:00   
39101          Столкновение  20180401   0:05:00   

                                             Причина ДТП  Число погибших  \
0                        Другие нарушения ПДД водителями               1   
1                        Другие нарушения ПДД водителями               1   
2             Превышение установленной скорости движения               1   
3      Нахождение на проезжей части без цели её перех...               0   
4          Нарушение правил проезда пешеходного перехода               1   
...               

### Преобразование типов данных

In [187]:
df = df.astype({
    'Код региона': 'category',
    'Регион': 'category',
    'Код дороги': 'category',
    'Название дороги': 'category',
    'Тип дороги': 'category',
    'Адрес': 'object',
    'Тип ДТП': 'category',
    'Причина ДТП': 'category',
    'Число погибших': 'object',
    'Число пострадавших': 'object',
    'Число транспортных средств': 'object',
    'Число участников': 'object'
})
df['Дата ДТП'] = pd.to_datetime(df['Дата ДТП'], format='%Y%m%d')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39102 entries, 0 to 39101
Data columns (total 14 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Код региона                 39102 non-null  category      
 1   Регион                      39102 non-null  category      
 2   Код дороги                  39102 non-null  category      
 3   Название дороги             39102 non-null  category      
 4   Тип дороги                  39102 non-null  category      
 5   Адрес                       39102 non-null  object        
 6   Тип ДТП                     39102 non-null  category      
 7   Дата ДТП                    39102 non-null  datetime64[ns]
 8   Время ДТП                   39102 non-null  object        
 9   Причина ДТП                 39102 non-null  category      
 10  Число погибших              39102 non-null  object        
 11  Число пострадавших          39102 non-null  object    

In [188]:
df.head()

Код региона                Регион   Код дороги  \
0          46    Московская область   Нет данных   
1          53  Оренбургская область  53К-2118000   
2          46    Московская область        А-107   
3           8      Хабаровский край        А-375   
4          35       Республика Крым      35Н-076   

                                     Название дороги  \
0                       мкр. Авиационный-Лямцино-ММК   
1                                 Обход г. Оренбурга   
2  "Московское малое кольцо" Икша - Ногинск - Бро...   
3  "Восток" Хабаровск - Красный Яр - Ариадное - Ч...   
4             Симферополь - Севастополь - Бахчисарай   

                                          Тип дороги  \
0  Региональная или межмуниципальная дорога общег...   
1  Региональная или межмуниципальная дорога общег...   
2              Федеральная дорога общего пользования   
3              Федеральная дорога общего пользования   
4  Региональная или межмуниципальная дорога общег...   

                                               Адрес               Тип ДТП  \
0  ЦФО, Московская область, Городской округ Домод...         Опрокидывание   
1                ПФО, Оренбургская область, Оренбург  Наезд на препятствие   
2         ЦФО, Московская область, Дмитровский район         Опрокидывание   
3           ДФО, Хабаровский край, Хабаровский район     Наезд на пешехода   
4         ЮФО, Республика Крым, Бахчисарайский район     Наезд на пешехода   

    Дата ДТП Время ДТП                                        Причина ДТП  \
0 2018-12-31  23:55:00                    Другие нарушения ПДД водителями   
1 2018-12-31  23:55:00                    Другие нарушения ПДД водителями   
2 2018-12-31  23:40:00         Превышение установленной скорости движения   
3 2018-12-31  23:20:00  Нахождение на проезжей части без цели её перех...   
4 2018-12-31  23:00:00      Нарушение правил проезда пешеходного перехода   

  Число погибших Число пострадавших Число транспортных средств  \
0              1                  0                          1   
1              1                  0                          1   
2              1                  0                          1   
3              0                  1                          1   
4              1                  0                          1   

  Число участников  
0               12  
1                2  
2                2  
3                2  
4                2

### Поиск дубликатов

In [189]:
df.duplicated().sum()


17

In [190]:
df.loc[df.duplicated(keep=False)]


Код региона                      Регион  Код дороги  \
37226          22       Нижегородская область    22К-0016   
37227          22       Нижегородская область    22К-0016   
39070          61           Рязанская область         М-5   
39071          46          Московская область  Нет данных   
39072          61           Рязанская область        Р-22   
39073          27     Калининградская область     27А-007   
39074          32         Кемеровская область       Р-255   
39075          61           Рязанская область       Р-132   
39076          86          Республика Карелия       А-121   
39077          56          Пензенская область         М-5   
39078          18       Волгоградская область        Р-22   
39079          46          Московская область  Нет данных   
39080          94       Удмуртская Республика         М-7   
39081          80     Республика Башкортостан     80Н-069   
39082          79  Республика Адыгея (Адыгея)  Нет данных   
39083           3          Краснодарский край       Р-217   
39084          76          Забайкальский край       А-350   
39085          63         Саратовская область  Нет данных   
39086          61           Рязанская область         М-5   
39087          46          Московская область  Нет данных   
39088          61           Рязанская область        Р-22   
39089          27     Калининградская область     27А-007   
39090          32         Кемеровская область       Р-255   
39091          61           Рязанская область       Р-132   
39092          86          Республика Карелия       А-121   
39093          56          Пензенская область         М-5   
39094          18       Волгоградская область        Р-22   
39095          46          Московская область  Нет данных   
39096          94       Удмуртская Республика         М-7   
39097          80     Республика Башкортостан     80Н-069   
39098          79  Республика Адыгея (Адыгея)  Нет данных   
39099           3          Краснодарский край       Р-217   
39100          76          Забайкальский край       А-350   
39101          63         Саратовская область  Нет данных   

                                         Название дороги  \
37226                                Семенов - Ковернино   
37227                                Семенов - Ковернино   
39070  "Урал" Москва - Рязань - Пенза - Самара - Уфа ...   
39071         МКАД-Крутицы-Железнодорожный-Ликино-Дулево   
39072  «Каспий» автомобильная дорога М-4 "Дон" - Тамб...   
39073                     Северный обход г. Калининграда   
39074                Подъезд к городу Томск от а/д Р-255   
39075                  Калуга - Тула - Михайлов - Рязань   
39076  "Сортавала" Санкт-Петербург - Сортавала - авто...   
39077  "Урал" Москва - Рязань - Пенза - Самара - Уфа ...   
39078  «Каспий» автомобильная дорога М-4 "Дон" - Тамб...   
39079      Куровское-Шатура-Дмитровский Погост-Самойлиха   
39080                  Подъезд к городу Пермь от а/д М-7   
39081                   Аскарово - Альмухаметово - Сибай   
39082                               Подъезд к а.Вочепший   
39083               Подъезд к городу Майкоп от а/д Р-217   
39084  Чита - Забайкальск - граница с Китайской Народ...   
39085  "автомобильная дорога ""Самара - Пугачев - Энг...   
39086  "Урал" Москва - Рязань - Пенза - Самара - Уфа ...   
39087         МКАД-Крутицы-Железнодорожный-Ликино-Дулево   
39088  «Каспий» автомобильная дорога М-4 "Дон" - Тамб...   
39089                     Северный обход г. Калининграда   
39090                Подъезд к городу Томск от а/д Р-255   
39091                  Калуга - Тула - Михайлов - Рязань   
39092  "Сортавала" Санкт-Петербург - Сортавала - авто...   
39093  "Урал" Москва - Рязань - Пенза - Самара - Уфа ...   
39094  «Каспий» автомобильная дорога М-4 "Дон" - Тамб...   
39095      Куровское-Шатура-Дмитровский Погост-Самойлиха   
39096                  Подъезд к городу Пермь от а/д М-7   
39097                   Аскарово - Альмухаметово - Сибай   
39098     

In [191]:
df.drop_duplicates(inplace=True)
df.duplicated().sum()

0

### Добавление столбцов.

In [192]:
# объединение столбцов crash_date и crash_time в один столбец datetime
df['datetime'] = pd.to_datetime(df['Дата ДТП'].astype(str) + 'T' + df['Время ДТП'].astype(str))

df = df.rename(columns={'datetime': 'Дата и время'})

In [193]:
# Добавление столбца с названиями дней недели на русском языке
df['День недели'] = df['Дата ДТП'].dt.day_name()


# Определение функции для замены названий дней недели
def translate_weekday(day):
    if day == 'Monday':
        return 'Понедельник'
    elif day == 'Tuesday':
        return 'Вторник'
    elif day == 'Wednesday':
        return 'Среда'
    elif day == 'Thursday':
        return 'Четверг'
    elif day == 'Friday':
        return 'Пятница'
    elif day == 'Saturday':
        return 'Суббота'
    elif day == 'Sunday':
        return 'Воскресенье'


# Замена английских названий дней недели на русские с помощью функции translate_weekday()
df['День недели'] = df['День недели'].apply(translate_weekday)
df

Код региона                      Регион   Код дороги  \
0              46          Московская область   Нет данных   
1              53        Оренбургская область  53К-2118000   
2              46          Московская область        А-107   
3               8            Хабаровский край        А-375   
4              35             Республика Крым      35Н-076   
...           ...                         ...          ...   
39081          80     Республика Башкортостан      80Н-069   
39082          79  Республика Адыгея (Адыгея)   Нет данных   
39083           3          Краснодарский край        Р-217   
39084          76          Забайкальский край        А-350   
39085          63         Саратовская область   Нет данных   

                                         Название дороги  \
0                           мкр. Авиационный-Лямцино-ММК   
1                                     Обход г. Оренбурга   
2      "Московское малое кольцо" Икша - Ногинск - Бро...   
3      "Восток" Хабаровск - Красный Яр - Ариадное - Ч...   
4                 Симферополь - Севастополь - Бахчисарай   
...                                                  ...   
39081                   Аскарово - Альмухаметово - Сибай   
39082                               Подъезд к а.Вочепший   
39083               Подъезд к городу Майкоп от а/д Р-217   
39084  Чита - Забайкальск - граница с Китайской Народ...   
39085  "автомобильная дорога ""Самара - Пугачев - Энг...   

                                              Тип дороги  \
0      Региональная или межмуниципальная дорога общег...   
1      Региональная или межмуниципальная дорога общег...   
2                  Федеральная дорога общего пользования   
3                  Федеральная дорога общего пользования   
4      Региональная или межмуниципальная дорога общег...   
...                                                  ...   
39081  Региональная или межмуниципальная дорога общег...   
39082  Региональная или межмуниципальная дорога общег...   
39083              Федеральная дорога общего пользования   
39084              Федеральная дорога общего пользования   
39085  Региональная или межмуниципальная дорога общег...   

                                                   Адрес  \
0      ЦФО, Московская область, Городской округ Домод...   
1                    ПФО, Оренбургская область, Оренбург   
2             ЦФО, Московская область, Дмитровский район   
3               ДФО, Хабаровский край, Хабаровский район   
4             ЮФО, Республика Крым, Бахчисарайский район   
...                                                  ...   
39081  ПФО, Республика Башкортостан, Абзелиловский ра...   
39082           ЮФО, Республика Адыгея, Теучежский район   
39083       ЮФО, Краснодарский край, Новокубанский район   
39084  СФО, Забайкальский край, Читинский район, Ново...   
39085        ПФО, Саратовская область, Энгельсский район   

                    Тип ДТП   Дата ДТП Время ДТП  \
0             Опрокидывание 2018-12-31  23:55:00   
1      Наезд на препятствие 2018-12-31  23:55:00   
2             Опрокидывание 2018-12-31  23:40:00   
3         Наезд на пешехода 2018-12-31  23:20:00   
4         Наезд на пешехода 2018-12-31  23:00:00   
...                     ...        ...       ...   
39081     Наезд на животное 2018-04-01   1:55:00   
39082     Наезд на животное 2018-04-01   1:30:00   
39083          Столкновение 2018-04-01   1:00:00   
39084         Опрокидывание 2018-04-01   0:30:00   
39085          Столкновение 2018-04-01   0:05:00   

                                             Причина ДТП Число погибших  \
0                        Другие нарушения ПДД водителями              1   
1                        Другие нарушения ПДД водителями              1   
2             Превышение установленной скорости движения              1   
3      Нахождение на проезжей части без цели её перех...              0   
4          Нарушение правил проезда пешеходного перехода              1   
...                    

### Экспорт результатов

In [196]:
df.to_excel('/Users/olgazavyalova/Documents/dtp_ready.xlsx', index=False)